tests/test_downloadermiddleware_cookies.py

```diff
@@ -732,3 +732,30 @@ def test_server_set_cookie_domain_public_period(self):
             "co.uk",
             cookies=True,
         )
+
+    def test_off_domain_jar_storage(self):
+        request1 = Request(
+            "https://a.example",
+            cookies=[
+                {
+                    "name": "foo",
+                    "value": "bar",
+                    "domain": "b.example",
+                },
+            ],
+        )
+        assert self.mw.process_request(request1, self.spider) is None
+        self.assertNotIn(b"Cookie", request1.headers)
+
+        request2 = Request("https://b.example/")
+        assert self.mw.process_request(request2, self.spider) is None
+        self.assertEqual(request2.headers.get(b"Cookie"), b"foo=bar")
+
+    def test_process_response_unrelated_domain(self):
+        request = Request("http://a.example")
+        request.headers["Cookie"] = "foo=bar; domain=b.example"
+        response = Response("http://c.example")
+        response.headers["Set-Cookie"] = "asd=fgh; domain=d.example"
+        self.mw.process_response(request, response, spider=None)
+        jar = self.mw.jars[None]
+        assert not jar._cookies.get("c.example", {}).get("/", {}).get("asd")
```

Comment:
The problem with the new approach is, I think, that it also applies to server-side cookies.

I think this test is not right, and fixing it as follows:

```suggestion
        response.headers["Set-Cookie"] = "asd=fgh; domain=c.example"
        self.mw.process_response(request, response, spider=None)
        jar = self.mw.jars[None]
        assert not jar._cookies.get("c.example", {}).get("/", {}).get("asd")
```

Will probably make it fail with the new approach.

It would not be very clean, but I wonder if it would be possible to change the policy right before setting user-defined cookies (and only user-defined cookies, not Set-Cookie headers from servers), and restore it right after.

If not that way, we need to think of something else.